talk/morelanguage/random.tex

```diff
@@ -0,0 +1,81 @@
+\subsection[a]{Random}
+%\epsdice[white]{3}
+
+\begin{frame}[fragile]
+  \frametitlecpp[11]{Pseudo-random number generators (PRNG)}
+  \begin{block}{Concept}
+    \begin{itemize}
+      \item For generating pseudo-random numbers the standard library offers:
+      \begin{itemize}
+        \item engines and engine adaptors to generate pseudo random bits
+        \item distributions to shape these into numbers
+        \item access to (potential) hardware entropy
+      \end{itemize}
+      \item Found in header \mintinline{cpp}{<random>}
+    \end{itemize}
+  \end{block}
+  \begin{cppcode}
+
+ \end{cppcode}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[11]{Engines}
+  \begin{block}{Engines}
+    \begin{itemize}
+      \item Generate random bits (as integers), depending on algorithm
+      \item Can be seeded via constructor or \mintinline{cpp}{seed()}
+      \item Algorithms: \mintinline{cpp}{linear_congruential_engine}, \mintinline{cpp}{mersenne_twister_engine}, \mintinline{cpp}{subtract_with_carry_engine}
+      \item Adaptors: \mintinline{cpp}{discard_block_engine}, \mintinline{cpp}{independent_bits_engine}, \mintinline{cpp}{shuffle_order_engine}
+      \item Aliases (use these): \mintinline{cpp}{minstd_rand0}, \mintinline{cpp}{minstd_rand}, \mintinline{cpp}{mt19937}, \mintinline{cpp}{mt19937_64}, \mintinline{cpp}{ranlux24}, \mintinline{cpp}{ranlux48}, \mintinline{cpp}{knuth_b}, \mintinline{cpp}{default_random_engine}
+    \end{itemize}
+  \end{block}
+  \begin{cppcode}
+    int seed = ...;
+    std::default_random_engine engine{seed};
+    auto i = engine(); // operator(): random integer
+ \end{cppcode}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[11]{Random device}
+  \begin{block}{Random device}
+    \begin{itemize}
+      \item PRNGs will produce the same numbers for the same seed
+      \item PRNGs are \emph{deterministic} (may be useful or not)
+      \item To get different numbers on each run, we must provide a truly random number as seed
+      \item \mintinline{cpp}{std::random_device} provides \emph{non-deterministic} random numbers
+      \item It uses hardware provided entropy, if available
+      \item It may become slow when called too often (e.g. when hardware entropy pool is exhausted)
+      \item Use it only to seed a PRNG!
+    \end{itemize}
+  \end{block}
+  \begin{cppcode}
+    std::random_device rd;
+    auto seed = rd();
+    std::default_random_engine engine{seed};
+ \end{cppcode}
```

Comment:
```suggestion
  \begin{block}{Random devices are deterministic}
    \begin{itemize}
      \item PRNGs will produce the same numbers for the same seed
      \item PRNGs are \emph{deterministic} (may be useful or not)
    \end{itemize}
  \end{block}
  \begin{block}{\texttt{std::random_device} provides \emph{non-deterministic} random numbers}
    \begin{itemize}
      \item It uses hardware provided entropy, if available
      \item It may become slow when called too often (e.g. when hardware entropy pool is exhausted)
      \item Use it only to seed a PRNG and get non-deterministic random numbers from it
    \end{itemize}
    \begin{cppcode}
      std::random_device rd;
      auto seed = rd();
      std::default_random_engine engine{seed};
   \end{cppcode}
  \end{block}
```

Reply:
I applied a variation of your suggestion.